sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Cast.java

```diff
@@ -0,0 +1,440 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.schemas.transforms;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Maps;
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.schemas.FieldAccessDescriptor;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.Field;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.beam.sdk.schemas.Schema.TypeName;
+import org.apache.beam.sdk.schemas.utils.SchemaZipFold;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+
+/** Set of utilities for casting rows between schemas. */
+@Experimental(Experimental.Kind.SCHEMAS)
+@AutoValue
+public abstract class Cast<T> extends PTransform<PCollection<T>, PCollection<Row>> {
+
+  public abstract Schema outputSchema();
+
+  public abstract Validator validator();
+
+  public static <T> Cast<T> of(Schema outputSchema, Validator validator) {
+    return new AutoValue_Cast<>(outputSchema, validator);
+  }
+
+  public static <T> Cast<T> widening(Schema outputSchema) {
+    return new AutoValue_Cast<>(outputSchema, Widening.of());
+  }
+
+  public static <T> Cast<T> narrowing(Schema outputSchema) {
+    return new AutoValue_Cast<>(outputSchema, Narrowing.of());
+  }
+
+  /** Describes compatibility errors during casting. */
+  @AutoValue
+  public abstract static class CompatibilityError implements Serializable {
+
+    public abstract List<String> path();
+
+    public abstract String message();
+
+    public static CompatibilityError create(List<String> path, String message) {
+      return new AutoValue_Cast_CompatibilityError(path, message);
+    }
+  }
+
+  /** Interface for statically validating casts. */
+  public interface Validator extends Serializable {
+    List<CompatibilityError> apply(Schema input, Schema output);
+  }
+
+  /**
+   * Widening changes to type that can represent any possible value of the original type.
+   *
+   * <p>Standard widening conversions:
+   *
+   * <ul>
+   *   <li>BYTE to INT16, INT32, INT64, FLOAT, DOUBLE, DECIMAL
+   *   <li>INT16 to INT32, INT64, FLOAT, DOUBLE, DECIMAL
+   *   <li>INT32 to INT64, FLOAT, DOUBLE, DECIMAL
+   *   <li>INT64 to FLOAT, DOUBLE, DECIMAL
+   *   <li>FLOAT to DOUBLE, DECIMAL
+   *   <li>DOUBLE to DECIMAL
+   * </ul>
+   *
+   * <p>Row widening:
+   *
+   * <ul>
+   *   <li>wider schema to schema with a subset of fields
```

Comment:
Doesn't match with the definition of widening, probably should be only in narrowing.